library/ziti.c

```diff
@@ -206,14 +212,72 @@ int ziti_init_opts(ziti_options *options, uv_loop_t *loop) {
     return ZITI_OK;
 }
 
-static void ziti_init_async(uv_async_t *ar) {
-    struct ziti_init_req *init_req = ar->data;
+extern bool ziti_is_enabled(ziti_context ztx) {
+    return ztx->enabled;
+}
 
-    ziti_context ztx = init_req->ztx;
-    uv_loop_t *loop = ar->loop;
-    ztx->id = ztx_seq++;
+extern void ziti_set_enabled(ziti_context ztx, bool enabled) {
+    ziti_queue_work(ztx, enabled ? ziti_start_internal : ziti_stop_internal, NULL);
+}
+
+static void logout_cb(void *resp, const ziti_error *err, void *ctx) {
+    ziti_context ztx = ctx;
+
+    free_ziti_session(ztx->session);
+    FREE(ztx->session);
+
+    model_map_clear(&ztx->sessions, (_free_f) free_ziti_net_session);
+    model_map_clear(&ztx->services, (_free_f) free_ziti_service);
+}
 
-    uv_close((uv_handle_t *) ar, (uv_close_cb) free);
+static void ziti_stop_internal(ziti_context ztx, void *data) {
+    if (ztx->enabled) {
+        ztx->enabled = false;
+
+        // stop updates
+        uv_timer_stop(&ztx->refresh_timer);
+        uv_timer_stop(&ztx->session_timer);
+        uv_timer_stop(ztx->posture_checks->timer);
+
+        // close all channels
+        ziti_close_channels(ztx, ZITI_DISABLED);
+
+        const char *svc_name;
+        ziti_service *svc;
+        ziti_event_t ev = {0};
+        ev.type = ZitiServiceEvent;
+        ev.event.service.removed = calloc(model_map_size(&ztx->services) + 1, sizeof(ziti_service *));
+        int idx = 0;
+        MODEL_MAP_FOREACH(svc_name, svc, &ztx->services) {
+            ev.event.service.removed[idx++] = svc;
+        }
+
+        ziti_send_event(ztx, &ev);
+        FREE(ev.event.service.removed);
+
+        // logout
+        ziti_ctrl_logout(&ztx->controller, logout_cb, ztx);
+
+        ev.type = ZitiContextEvent;
+        ev.event.ctx.ctrl_status = ZITI_DISABLED;
+
+        ziti_send_event(ztx, &ev);
+    }
+}
+
+static void ziti_start_internal(ziti_context ztx, void *init_req) {
+    if (!ztx->enabled) {
```

Comment:
personally, i think this could use some log statements. this might be hard to debug/diagnose without.